### PR TITLE
Add --format support to images command

### LINKED ISSUE #18360: Enhancement: docker images --format
Would be great to have a similar one to [docker ps --format](https://docs.docker.com/engine/reference/commandline/ps/#formatting) option.

### PR DESCRIPTION
**Pull request by vdemeester**
closes https://github.com/docker/docker/issues/18360

The `ps` commands had `--format` flags, this add it to the `images` command. This kinda fixes issues like : #17304. 🐮
- It refactors and rename the package `api/client/ps` to `api/client/formatter`,
- It adds a a image formatter and update `api/client/images.go`.

The default output of `docker images` remain unchanged. But it's now possible to use a `--format` flag and to setup a default `images` format in the client configuration:

``` bash
$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED AT                          VIRTUAL SIZE
swarm               latest              6b40fe7724bd        2015-11-03 20:07:32 +0000 UTC ago   15.6 MB
nginx               latest              914c82c5a678        2015-10-28 22:01:52 +0000 UTC ago   132.7 MB

$ docker images --format="{{.ID}}"
6b40fe7724bd
914c82c5a678

$ docker images --format="table {{.ID}}\t{{.Repository}}"
IMAGE ID            REPOSITORY
6b40fe7724bd        swarm
914c82c5a678        nginx

$ docker images --no-trunc --format="{{.ID}}\t{{.Repository}}"
6b40fe7724bd29107f6182ca2befec011cdf524b23ebc4c9a33591d6b7aea4ee        swarm
914c82c5a678cf9fa73c25d91d30f5e174ac69f6b494890d262cc58f43b304ef        nginx

$ docker images --format="raw"
repository: swarm
tag: latest
image_id: 6b40fe7724bd
created_at: 2015-11-03 20:07:32 +0000 UTC
virtual_size: 15.6 MB
repository: nginx
tag: latest
image_id: 914c82c5a678
created_at: 2015-10-28 22:01:52 +0000 UTC
virtual_size: 132.7 MB
```

What still need to do :
- [X] Update/Rewrite the tests for `api/client/formatter`
- [X] Add integration tests
- [X] Refactor a bit the code, as it is not that great as of now, there is room for improvement :stuck_out_tongue_closed_eyes:.
- [X] update shell completion
- [X] Update docs :wink: 

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

**Comment by albers** [comment]
I like the feature.

@vdemeester Please add the completions to the todo list. Unless you prefer to implement it yourself, I will take over the bash completion part.

**Comment by vdemeester** [comment]
@albers ah true, I'll add it to my list of things to do :wink:

**Comment by vdemeester** [comment]
oh there is a test failing in `DockerRegistrySuite` I didn't spot locally :sweat_smile:

**Comment by calavera** [comment]
Awesome, one thing less in my very long queue :smile_cat: 

I think @vieux also wanted this.

**Comment by LK4D4** [comment]
@vdemeester pls add one-two integration tests.

**Comment by vdemeester** [comment]
I will :wink: during DockerCon or just after :wink:

File: contrib/completion/bash/docker [hunk 1]
```diff
@@ -933,7 +933,7 @@ _docker_images() {
 
 	case "$cur" in
 		-*)
-			COMPREPLY=( $( compgen -W "--all -a --digests --filter -f --help --no-trunc --quiet -q" -- "$cur" ) )
+			COMPREPLY=( $( compgen -W "--all -a --digests --filter -f --format --help --no-trunc --quiet -q" -- "$cur" ) )
```
Comment (albers) [inline contrib/completion/bash/docker]: As `--format` takes an argument, completion also has to take care of that, see https://github.com/docker/docker/blob/master/contrib/completion/bash/docker#L984-L987 for an example.

Comment (vdemeester) [inline contrib/completion/bash/docker]: @albers hehe, same as network `--format` :sweat_smile:

File: api/client/formatter/formatter.go [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+virtual_size: {{.VirtualSize}}`
+			} else {
+				ctx.Format = `repository: {{ .Repository }}
+tag: {{.Tag}}
+image_id: {{.ID}}
+created_at: {{.CreatedAt}}
+virtual_size: {{.VirtualSize}}`
+			}
+		}
+	default:
```
Comment (calavera) [inline api/client/formatter/formatter.go]: this default key is not necessary.

File: api/client/formatter/formatter.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+names: {{.Names}}
+labels: {{.Labels}}
+ports: {{.Ports}}
+`
+			if ctx.Size {
+				ctx.Format += `size: {{.Size}}
+`
+			}
+		}
+	default:
```
Comment (calavera) [inline api/client/formatter/formatter.go]: this default key is not necessary.

File: api/client/formatter/formatter.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+created_at: {{.CreatedAt}}
+virtual_size: {{.VirtualSize}}`
+			}
+		}
+	default:
+	}
+	customImageFormat(ctx)
+}
+
+func customContainerFormat(ctx ContainerContext) {
```
Comment (calavera) [inline api/client/formatter/formatter.go]: there is a lot of duplication between this function and `customImagesFormat`. Is there a way to reduce it?

**Comment by calavera** [comment]
This is going to need documentation, but it's looking great :raised_hands:

File: api/client/formatter/formatter.go (hunk 4 above)
Comment (vdemeester) [inline api/client/formatter/formatter.go]: Yeah it's on my todo up there… I'm looking for a way to reduce it (just didn't got time to work on that :sweat_smile:)

**Comment by thaJeztah** [comment]
ping @calavera looks like code was updated, PTAL

**Comment by calavera** [comment]
LGTM

**Comment by thaJeztah** [comment]
ping @duglin perhaps you can have a look at this one? :heart:

**Comment by vdemeester** [comment]
Rebased and added simple integration tests.

**Comment by vdemeester** [comment]
ping @LK4D4 @runcom @duglin for reviews :stuck_out_tongue_closed_eyes:

**Comment by thaJeztah** [comment]
looks like this needs a rebase again @vdemeester, I'll start harassing some maintainers after that :wink:

**Comment by vdemeester** [comment]
Done :sweat_smile:

**Comment by thaJeztah** [comment]
docs LGTM (@LK4D4 wanted to merge sooooo badly, haha)

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**